crates/phactory/src/contracts/support.rs

```diff
@@ -93,105 +62,62 @@ pub trait NativeContract {
         Self: Sized;
 }
 
-#[derive(Serialize, Deserialize, Encode, Decode)]
-pub struct NativeContractWrapper<Con> {
-    inner: Con,
-    id: sp_core::H256,
-}
-
-impl<Con> NativeContractWrapper<Con> {
-    pub fn new(inner: Con, contract_info: &ContractInfo<chain::Hash, chain::AccountId>) -> Self {
-        let id = contract_info.contract_id(blake2_256);
-        NativeContractWrapper { inner, id }
-    }
+pub(crate) struct Query {
+    contract: AnyContract,
 }
 
-impl<Con: NativeContract> NativeContract for NativeContractWrapper<Con> {
-    type Cmd = Con::Cmd;
-    type QReq = Con::QReq;
-    type QResp = Con::QResp;
-
-    fn handle_command(
-        &mut self,
-        origin: MessageOrigin,
-        cmd: Self::Cmd,
-        context: &mut NativeContext,
-    ) -> TransactionResult {
-        self.inner.handle_command(origin, cmd, context)
-    }
-
-    fn handle_query(
+impl Query {
+    pub fn handle_query(
         &self,
         origin: Option<&runtime::AccountId>,
-        req: Self::QReq,
+        req: OpaqueQuery,
         context: &mut QueryContext,
-    ) -> Self::QResp {
-        self.inner.handle_query(origin, req, context)
-    }
-
-    fn on_block_end(&mut self, context: &mut NativeContext) -> TransactionResult {
-        self.inner.on_block_end(context)
-    }
-
-    fn snapshot(&self) -> Self {
-        Self {
-            inner: self.inner.snapshot(),
-            id: self.id,
-        }
+    ) -> Result<OpaqueReply, OpaqueError> {
+        self.contract.handle_query(origin, req, context)
     }
 }
 
-impl<Con: NativeContract> NativeContractMore for NativeContractWrapper<Con> {
-    fn id(&self) -> phala_mq::ContractId {
-        self.id
-    }
-}
+pub(crate) struct RawData(Vec<u8>);
 
-struct Query<Con> {
-    contract: Con,
-}
-
-impl<Con> Queryable for Query<Con>
-where
-    Con: NativeContract,
-{
-    fn handle_query(
-        &self,
-        origin: Option<&runtime::AccountId>,
-        req: OpaqueQuery,
-        context: &mut QueryContext,
-    ) -> Result<OpaqueReply, OpaqueError> {
-        let response = self
-            .contract
-            .handle_query(origin, deopaque_query(req)?, context);
-        Ok(response.encode())
+impl Decode for RawData {
+    fn decode<I: parity_scale_codec::Input>(
+        input: &mut I,
+    ) -> Result<Self, parity_scale_codec::Error> {
+        let capacity = input.remaining_len()?.unwrap_or_default();
+        let mut buf = Vec::with_capacity(capacity);
+        loop {
+            match input.read_byte() {
```

Comment:
Better to use [read()](https://paritytech.github.io/substrate/master/parity_scale_codec/trait.Input.html#tymethod.read) to batch the read?

Reply:
The `Input::read` is different from the std `Read::read`. It would return Err if there are no enough data to fill the given buffer. So it would be hard to known how many bytes read in the last batch.

Reply:
Got it. Thx for the explaination